lib/IRGen/GenBuiltin.cpp

```diff
@@ -1543,3 +1545,371 @@ void irgen::emitBuiltinCall(IRGenFunction &IGF, const BuiltinInfo &Builtin,
+                                Element.getFixedAlignment().getValue(), *this);
+  }
+
+  void loadAsCopy(IRGenFunction &IGF, Address addr,
+                  Explosion &explosion) const override {
+    eachElementAddr(IGF, addr,
+                    [&](Address elementAddr) {
+                      Element.loadAsCopy(IGF, elementAddr, explosion);
+                    });
+  }
+
+  void loadAsTake(IRGenFunction &IGF, Address addr,
+                  Explosion &explosion) const override {
+    eachElementAddr(IGF, addr,
+                    [&](Address elementAddr) {
+                      Element.loadAsTake(IGF, elementAddr, explosion);
+                    });
+  }
+  
+  void assign(IRGenFunction &IGF, Explosion &explosion, Address addr,
+              bool isOutlined, SILType T) const override {
+    auto eltTy = getElementSILType(IGF.IGM, T);
+    eachElementAddr(IGF, addr,
+                    [&](Address elementAddr) {
+                      Element.assign(IGF, explosion, addr, isOutlined,
+                                     eltTy);
+                    });
+  }
+
+  void initialize(IRGenFunction &IGF, Explosion &explosion, Address addr,
+                  bool isOutlined) const override {
+    eachElementAddr(IGF, addr,
+                    [&](Address elementAddr) {
+                      Element.initialize(IGF, explosion, addr, isOutlined);
+                    });
+  }
+  
+  void reexplode(Explosion &sourceExplosion,
+                 Explosion &targetExplosion) const override {
+    eachElement([&]{
+      Element.reexplode(sourceExplosion, targetExplosion);
+    });
+  }
+  
+  void copy(IRGenFunction &IGF,
+            Explosion &sourceExplosion,
+            Explosion &targetExplosion,
+            Atomicity atomicity) const override {
+    eachElement([&]{
+      Element.copy(IGF, sourceExplosion, targetExplosion, atomicity);
+    });
+  }
+  
+  void consume(IRGenFunction &IGF, Explosion &explosion,
+               Atomicity atomicity,
+               SILType T) const override {
+    auto eltTy = getElementSILType(IGF.IGM, T);
+    eachElement([&]{
+      Element.consume(IGF, explosion, atomicity, eltTy);
+    });
+  }
+
+  void initializeFromParams(IRGenFunction &IGF, Explosion &params,
+                            Address src, SILType T,
+                            bool isOutlined) const override {
+    auto eltTy = getElementSILType(IGF.IGM, T);
+    eachElementAddr(IGF, src,
+                    [&](Address elementAddr) {
+                      Element.initializeFromParams(IGF, params, elementAddr,
+                                                   eltTy, isOutlined);
+                    });
+  }
+  
+  void fixLifetime(IRGenFunction &IGF,
+                   Explosion &explosion) const override {
+    eachElement([&]{
+      Element.fixLifetime(IGF, explosion);
+    });
+  }
+  
+  template<typename...Addresses, typename Body>
+  void eachElementAddrLoop(IRGenFunction &IGF,
```

Comment:
Not sure how useful it'd be here, but I added `IRGenFunction::emitLoopOverElements` to support the `ArrayLayoutEntry` work.

Reply:
Ah, I see! I'll work on reconciling these in a follow up since it would be good to have only one implementation. `emitLoopOverElements` looks like it has some bugs in its ad-hoc element indexing that could be avoided by using `TypeInfo::indexArray`.